pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+			Groups::<T>::try_mutate(group_id, |group| {
+				group.distribute_reward(reward)?;
+
+				T::Currency::mint_into(
+					T::RewardCurrency::get(),
+					&T::PalletId::get().into_account_truncating(),
+					reward,
+				)?;
+
+				Self::deposit_event(Event::GroupRewarded {
+					group_id,
+					amount: reward,
+				});
+
+				Ok(())
+			})
+		}
+
+		fn group_stake(group_id: Self::GroupId) -> Self::Balance {
+			Groups::<T>::get(group_id).total_staked()
+		}
+	}
+
+	impl<T: Config> AccountRewards<T::AccountId> for Pallet<T>
+	where
+		T::Balance: EnsureAdd + EnsureSub,
+	{
+		type Balance = T::Balance;
+		type CurrencyId = T::CurrencyId;
+
+		fn deposit_stake(
+			currency_id: Self::CurrencyId,
+			account_id: &T::AccountId,
+			amount: Self::Balance,
+		) -> DispatchResult {
+			Currencies::<T>::try_mutate(currency_id, |currency| {
+				let group_id = currency.group_id.ok_or(Error::<T>::CurrencyWithoutGroup)?;
+
+				Groups::<T>::try_mutate(group_id, |group| {
+					StakeAccounts::<T>::try_mutate(account_id, currency_id, |staked| {
+						T::Currency::hold(currency_id, &account_id, amount)?;
+
+						staked.try_apply_rpt_tallies(currency.rpt_tallies())?;
+						staked.add_amount(amount, group.reward_per_token())?;
+
+						group.add_amount(amount)?;
+						currency.add_amount(amount)?;
+
+						Self::deposit_event(Event::StakeDeposited {
+							group_id,
+							currency_id,
+							account_id: account_id.clone(),
+							amount,
+						});
+
+						Ok(())
+					})
+				})
+			})
+		}
+
+		fn withdraw_stake(
+			currency_id: Self::CurrencyId,
+			account_id: &T::AccountId,
+			amount: Self::Balance,
+		) -> DispatchResult {
+			Currencies::<T>::try_mutate(currency_id, |currency| {
+				let group_id = currency.group_id.ok_or(Error::<T>::CurrencyWithoutGroup)?;
+
+				Groups::<T>::try_mutate(group_id, |group| {
+					StakeAccounts::<T>::try_mutate(account_id, currency_id, |staked| {
+						T::Currency::release(currency_id, &account_id, amount, false)?;
+
+						staked.try_apply_rpt_tallies(currency.rpt_tallies())?;
+						staked.sub_amount(amount, group.reward_per_token())?;
+
+						group.sub_amount(amount)?;
+						currency.sub_amount(amount)?;
+
+						Self::deposit_event(Event::StakeWithdrawn {
+							group_id,
+							currency_id,
+							account_id: account_id.clone(),
+							amount,
+						});
+
+						Ok(())
+					})
+				})
+			})
+		}
+
+		fn compute_reward(
+			currency_id: Self::CurrencyId,
+			account_id: &T::AccountId,
+		) -> Result<Self::Balance, DispatchError> {
+			let currency = Currencies::<T>::get(currency_id);
+			let group_id = currency.group_id.ok_or(Error::<T>::CurrencyWithoutGroup)?;
+			let group = Groups::<T>::get(group_id);
+
+			StakeAccounts::<T>::try_mutate(account_id, currency_id, |staked| {
+				staked.try_apply_rpt_tallies(currency.rpt_tallies())?;
+				let reward = staked.compute_reward(group.reward_per_token())?;
+
+				Ok(reward)
+			})
+		}
+
+		fn claim_reward(
```

Comment:
Would be great if we could add another function that claims for all the currencies that the account staked on. Not a hard requirement, but much better UX since it will only require one call instead of n depending on number of currencies staked. 
We can as well add it later.

Reply:
Seems a good idea! Although we will need a new storage `Map<AccountId, vec<CurrencyId>>` to know which currencies are needed to check for a specific account.
Not sure if this logic should be in `pallet-rewards` or `pallet-liquidity-rewards`. I think the last one.

Reply:
I have been following the development of this pallet and wanted to share my two cents:

IMO, such functionality should rather be handled in the Frontend than be part of the runtime. You can expose the list of staked (and claimable) currencies via a runtime API call and then claim these in a batch call. However, if it should be easily usable in Polkadot Js, an extrinsic would be helpful.

For the mapping of `AccountId` to staked `CurrencyId`s you can make use of [`iter_key_prefix(account_id)`](https://paritytech.github.io/substrate/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.iter_key_prefix). IMO, it would be helpful to have a runtime API call exposing all staked currencies and another one filtering only for those for which you can claim.

Reply:
Thanks for the comments @wischli 

> via a runtime API call and then claim these in a batch call

Good idea 🚀 

> you can make use of [StakeAccounts::<T>::iter_key_prefix(account_id)](https://paritytech.github.io/substrate/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.iter_key_prefix)

I like it too. My only concern is that this pallet can be used for any other number of pallets. Some currencies can be used in one pallet, and others currencies in another pallet. Maybe the user is only interested in claiming the currencies of one of those pallets. Anyway, it seems like a very clean solution that at least we should add to the [`Rewards`](https://github.com/centrifuge/centrifuge-chain/pull/1036) API.

Reply:
I.e. I would strongly lobby for single claim logic per currency and let the front combine as wished via batch calls.

Reply:
> we can provide a separate runtime-api and rpc-call for this purpose too

It has a lot of sense 👍🏻

[`StakeAccounts::::iter_key_prefix(account_id)`](https://paritytech.github.io/substrate/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.iter_key_prefix) maybe has sense for the first iteration of CFG Rewards. With more time, we can make updates to make it by rpc calls once we use several pallets or our currencies start to grow in number.

Reply:
Thanks @wischli for the comments. 
To add to what @mustermeiszer & @lemunozm mentioned. As a user I would like to:
- Get a list of currencies that I have staked in. (Runtime API)
- Have a way to see my **total** claimable rewards across all currencies I invested in. (Runtime API)
- Be able to claim rewards for one or many currencies. (Combination of Runtime API + extrinsic/batch extrinsic)

The first two I think are straight forward using [StakeAccounts::::iter_key_prefix(account_id)](https://paritytech.github.io/substrate/master/frame_support/storage/trait.IterableStorageDoubleMap.html#tymethod.iter_key_prefix) in the API level. 
For the third one, @mustermeiszer if we define a function that takes a Vec<CurrencyId, Max(20?)> then we can achieve both for most of the users. And if some users need to repeat the call because they staked in more than 20? currencies then use the Utility.Batch() call. The user/api would have to first call the `Runtime API, to get the total list of currencies` and then decide how to batch it, but at least for most of the users there will be no need to do a Batch call. 
Of course the benchmark here is important as always.

Reply:
@mustermeiszer sounds good, lets go for that.